### ISSUE TITLE
OpenJDK is somewhat broken on newer MacOS instances, console is flooded with errors when using JMeter, AdoptOpenJDK has no issues

### ISSUE THREAD

**Issue by Dosiara** [Homebrew/homebrew-core#66953]
# Bug report

- [X] ran `brew update` and can still reproduce the problem?
- [X] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [X] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
`No Logs.`
- [X] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

```
HOMEBREW_VERSION: 2.6.2
ORIGIN: https://github.com/Homebrew/brew
HEAD: 1a52862e6d24cd3bf87bec557a8886a1a825ee3c
Last commit: 26 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: d914aab917710e94170526641d16414cb6a2083e
Core tap last commit: 21 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit westmere
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
Java: 15.0.1, 15
macOS: 11.1-x86_64
CLT: 12.3.0.0.1.1607026830
Xcode: 12.3
Rosetta 2: true
```

## What you were trying to do (and why)

```
Was trying to run Jmeter on MacOS 11.1 like a normal load-testing being.
Got flooded with the following errors making it hard to spot actual problems.
```

## What happened (include command output)

```
================================================================================
Don't use GUI mode for load testing !, only for Test creation and Test debugging.
For load testing, use CLI Mode (was NON GUI):
   jmeter -n -t [jmx file] -l [results file] -e -o [Path to web report folder]
& increase Java Heap to meet your test requirements:
   Modify current env variable HEAP="-Xms1g -Xmx1g -XX:MaxMetaspaceSize=256m" in the jmeter batch file
Check : https://jmeter.apache.org/usermanual/best-practices.html
================================================================================
2020-12-15 18:15:33.705 java[11476:123425] JavaNativeFoundation: GetGlobalVM: Failed to locate @rpath/libjvm.dylib for JNI_GetCreatedJavaVMs(). A JVM must be loaded before calling this function.
2020-12-15 18:15:33.918 java[11476:123425] JavaNativeFoundation: GetGlobalVM: Failed to locate @rpath/libjvm.dylib for JNI_GetCreatedJavaVMs(). A JVM must be loaded before calling this function.
```

## What you expected to happen

`There should have been no Failed to locate @rpath/libjvm.dylib errors en masse.`

## Step-by-step reproduction instructions (by running `brew install` commands)

```
brew install jmeter
jmeter
```

**Comment by Dosiara** [comment]
First time submitter, I might have been wrong on making an issue in here. Please point me to the correct place before closing if that's the case.

**Comment by SMillerDev** [comment]
Is there a specific reason you think this is caused by homebrew?

**Comment by Dosiara** [comment]
> Is there a specific reason you think this is caused by homebrew?

Well, it is caused be a specific brew cask that lists OpenJDK as dependency and links Jmeter execution to bin-bash that specifically calls OpenJDK by modifying original Jmeter package. It could be solved either by changing Jmeter package dependencies and launch script or by changing something in the OpenJDK package: whether it is brew related or OpenJDK is at fault -- I do not know yet.
The easiest solution would be to change installation script, specifically /bin/jmeter exec to use AdoptOpenJDK, or prebuilt JVM. It works just fine.

This behaviour has been reproduced at other machines as well.

Whom should I contact in this regard? What is my best course of action?

**Comment by carlocab** [comment]
>What is my best course of action?

The best is probably to work out a way to implement one of the fixes you've suggested and to open a PR.

**Comment by Dosiara** [comment]
> > What is my best course of action?
> 
> The best is probably to work out a way to implement one of the fixes you've suggested and to open a PR.

That seem to be the case. My solution is somewhat ugly, since it switches from source based formula to a cask. I will try with OpenJDK built from source and will proceed from that.

**Comment by carlocab** [comment]
>I will try with OpenJDK built from source and will proceed from that.

This sounds ideal. Unless I've misunderstood something, you may even be able to use the bottled `openjdk`.

**Comment by Dosiara** [comment]
> > I will try with OpenJDK built from source and will proceed from that.
> 
> This sounds ideal. Unless I've misunderstood something, you may even be able to use the bottled `openjdk`.

Nah, the bottled version has his problem. There was an issue opened and closed, but the problem remains. https://github.com/CoatiSoftware/Sourcetrail/issues/757

**Comment by Dosiara** [comment]
> > I will try with OpenJDK built from source and will proceed from that.
> 
> This sounds ideal. Unless I've misunderstood something, you may even be able to use the bottled `openjdk`.

It is stranger than I thought. While bundled java 15.0.1 and `adoptopenjdk` both have the new location of `libjvm.dylib` (as per the issue above), `openjdk` has none at all. It actually lacks the majority of structure present in other 2 implementations. Hows that work?

Built from scrath `openjdk` behaves the same way as the bottled one for me.

It leaves me with the only option to PR a change of dependencies it seems..

**Comment by carlocab** [comment]
In my install of `openjdk`, I found `libjvm.dylib` at `/usr/local/opt/openjdk/libexec/openjdk.jdk/Contents/Home/lib/server/libjvm.dylib`. Perhaps it's just an issue of setting the right compiler flags?

**Comment by Dosiara** [comment]
> In my install of `openjdk`, I found `libjvm.dylib` at
> 
> ```
> /usr/local/opt/openjdk/libexec/openjdk.jdk/Contents/Home/lib/server/libjvm.dylib
> ```
> 
> Perhaps it's just an issue of setting the right compiler flags?

I was not aware that we can go into the package. So the lib is in place, thank you.

Just now I was told that the issue is not repeating itself when `OpenJDK` is not from Brew. Will report back.

**Comment by Dosiara** [comment]
So the problem IS with Brew's `openjdk` package and not with OpenJDK per se, or `jmeter` package.
OpenJDK tarball from https://jdk.java.net/15/ works just fine!

Where do I go from that? The problem is clearly with whatever Brew does to OpenJDK.

**Comment by SMillerDev** [comment]
Since it's not in adoptopenjdk for example you could check how they do their build and try and spot where brew differs on the libjvm part.

**Comment by bayandin** [comment]
It seems, that this issue affects only Big Sur (it doesn't produce such kind of error on Catalina)

**Comment by Dosiara** [comment]
I think I narrowed the problem down to the contents of `openjdk.jdk`. If they are replace with java.net's tarball everything becomes fine.
The first suspects were missing `CodeResources` file and `_CodeSignature` folder, but their addition does nothing.
There is a minuscule difference in `Info.plist` file that boils down to spaces and version numbers.

So it is either the patches or build flags. Am I on a right track?

**Comment by Dosiara** [comment]
It might be the dynamic library linking introduced in 11.0.1 [https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-release-notes](url) requiring calling libs with `dlopen()`. I've looked at `OpenJDK` source code used in the Brew `openjdk` formula for bottling and it does not seem to obey the rule.
On the other hand, java.net's build works just fine. What do I make from that?

**Comment by carlocab** [comment]
I'm not sure that's related; I believe the need to use `dlopen` applies only to system libraries. You didn't need to do this previously since system libraries used to also be present in the file system at `/usr/lib`. The missing libraries from the `openjdk` formula are not system libraries.

**Comment by tresf** [comment]
If you modify the formula to explicitly create the expected symlink, does it fix it?  I had to do [something similar getting the Linux formula to work](https://github.com/Homebrew/homebrew-core/blob/76b5fdef7cfe963c73843d56193f79c0843346b5/Formula/openjdk.rb#L115-L119).

**Comment by jonchang** [comment]
I look a peek at how adoptopenjdk was building their openjdk binaries. Their configure flags look like

```
--disable-warnings-as-errors
--enable-ccache
--enable-dtrace
--verbose
--with-boot-jdk=/Users/jenkins/workspace/build-scripts/jobs/jdk15u/jdk15u-mac-x64-hotspot/jdk-14/Contents/Home
--with-cacerts-file=/Users/jenkins/workspace/build-scripts/jobs/jdk15u/jdk15u-mac-x64-hotspot/sbin/../security/cacerts
--with-debug-level=release
--with-extra-cxxflags=-mmacosx-version-min=10.9
--with-freetype=bundled
--with-jvm-variants=server
--with-macosx-codesign-identity='Developer ID Application: London Jamocha Community CIC'
--with-native-debug-symbols=none
--with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/
--with-vendor-bug-url=https://github.com/AdoptOpenJDK/openjdk-support/issues
--with-vendor-name=AdoptOpenJDK
--with-vendor-url=https://adoptopenjdk.net/
--with-vendor-version-string=AdoptOpenJDK
--with-vendor-vm-bug-url=https://github.com/AdoptOpenJDK/openjdk-support/issues
--with-version-build=9
--with-version-opt=202012282111
--without-version-pre
```

whereas ours (for openjdk 15) are:

```
--enable-dtrace
--with-boot-jdk-jvmargs=#{java_options}
--with-boot-jdk=#{boot_jdk}
--with-debug-level=release
--with-extra-ldflags=-headerpad_max_install_names
--with-jvm-variants=server
--with-native-debug-symbols=none
--with-sysroot=#{MacOS.sdk_path}
--with-toolchain-path=/usr/bin
--with-version-build=#{build}
--without-version-opt
--without-version-pre
```

There's nothing suspicious in the configure flags as the adoptopenjdk configuration is for patch builds, and we should probably set the vendor stuff at some point too actually. If we `brew install --cask adoptopenjdk` and examine libzip.dylib the difference is apparent:

adoptopenjdk:

```console
% otool -L libzip.dylib
libzip.dylib:
	@rpath/libzip.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libjvm.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libjava.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```

brewed openjdk:

```console
% otool -L libzip.dylib
libzip.dylib:
	/usr/local/opt/openjdk/libexec/openjdk.jdk/Contents/Home/lib/libzip.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libjvm.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libjava.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
```

There's definitely something funny going on. We may need to investigate if `keg_relocate` is doing something weird to the dylib names.

**Comment by ipstone** [comment]
I have the following error on Catalina:

Error: Failed to download resource "openjdk_bottle_manifest"
Download failed: https://ghcr.io/v2/homebrew/core/openjdk/manifests/13.0.2.8

**Comment by carlocab** [comment]
> We may need to investigate if `keg_relocate` is doing something weird to the dylib names.

The difference in install name comes from [`fix_dynamic_linkage`](https://github.com/Homebrew/brew/blob/f9dd12ef5c4cc4bd8ac5225fea78d6a30848938c/Library/Homebrew/extend/os/mac/keg_relocate.rb#L44) in `extend/os/mac/keg_relocate.rb`.

We tried not changing dylib IDs that are prefixed with `@rpath` before (see https://github.com/Homebrew/brew/pull/2036) as a way of avoiding this workaround:

https://github.com/Homebrew/homebrew-core/blob/dc41af4d64bb9a9dcc8956e663ddc9c7d54f8583/Formula/rust.rb#L110-L116

But that was reverted, since it made linking to libraries with `@rpath` in their ID annoying.

**Comment by ajit-y** [comment]
I was getting following error as well after upgrade to Bigsur
`JavaNativeFoundation: GetGlobalVM: Failed to locate @rpath/libjvm.dylib for JNI_GetCreatedJavaVMs(). A JVM must be loaded before calling this function.`
I used this workaround to resolve the issue.
1- I am using jnev to have and manage multiple versions of java on Mac. So I have installed adoptopenjdk after reading the comments here that adoptopenjdk soes not have this issue.
2- I navigated to file /usr/local/Cellar/jmeter/5.4.1/bin/jmeter and updated it as below

`#!/bin/bash
#JAVA_HOME="/usr/local/opt/openjdk" exec "/usr/local/Cellar/jmeter/5.4.1/libexec/bin/jmeter"  "$@"
JAVA_HOME="/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home" exec "/usr/local/Cellar/jmeter/5.4.1/libexec/bin/jmeter"  "$@"` 

so basically I have just provided path to adoptopenjdk in 'JAVA_HOME' and commented the original path which could be just removed as well.

I stopped getting that error after this and able to use jmeter.

**Comment by tresf** [comment]
> so basically I have just provided path to adoptopenjdk in 'JAVA_HOME' and commented the original path which could be just removed as well.

Slight warning for M1 users, AdoptOpenJDK doesn't yet provide an M1-native release, so this workaround will falter for aarch64.  Note also, `/usr/local` is `/opt/local` on M1 native installs of Homebrew.

**Comment by tudorsss** [comment]
Hi @ajit-y ,

Could you please walk us though a bit more from the beginning, what needs to be done, in order for us to apply the workaround?

I have not used jenv before, so I don't know how to use it and how to set it up.

Thanks!

**Comment by ajit-y** [comment]
Hi @tudorsss ,

I am wondering if the mention of jenv in my comment just unnecessary added the complication to work around. Because I really didn't do anything with jenv while working this around. I do use Jenv for managing multiple jdks on my machine and its very good, its detailed documentation is at https://www.jenv.be/ if you are interested.

Could you try -
1- Install adoptopenjdk using brew. I have adoptjopendk11 on my system so I think I used "brew install --cask adoptopenjdk11" to install it.
2- Find the directory where jdk is installed. I could find it on my machine at /Library/Java/JavaVirtualMachines
3- Find the path to Home directory for the jdk. I could find it at /Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home
4- I navigated to directory /usr/local/Cellar/jmeter/5.4.1/bin   . Note 5.4.1 is the version of jmeter on my machine.
5- Edit file jmeter

Remove or comment line -
#!/bin/bash #JAVA_HOME="/usr/local/opt/openjdk" exec "/usr/local/Cellar/jmeter/5.4.1/libexec/bin/jmeter" "$@" 

Add line -
JAVA_HOME="/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home" exec "/usr/local/Cellar/jmeter/5.4.1/libexec/bin/jmeter" "$@"

Again, please be careful with version numbers you are using for jdk and jmeter.

Give it a try. Hope it should work.
Hope you also have noted comment from @tresf regarding M1 users.

**Comment by tresf** [comment]
> Hope you also have noted comment from @tresf regarding M1 users.

Hmm... it seems that the issue doesn't occur on M1 builds.  Performing the following:

> brew install jmeter
> jmeter

... I do NOT receive the `Failed to locate @rpath/libjvm.dylib` errors.  This could be environmental, but I don't think so.... Reading the jmeter formula as well as the `/opt/homebrew/bin/jmeter` script, it most certainly is setting `JAVA_HOME` properly, so this may be a slight difference between formulas, perhaps even a regression introduced in some recent changes to the [`openjdk` formula](https://github.com/Homebrew/homebrew-core/commits/8454896e7ab23b37f73d29f97a73ae897a075a91/Formula/openjdk.rb).

In regards to the error itself, `Failed to locate @rpath/libjvm.dylib`, `@rpath` is intended to be relative to the binary location.  This means at some point an executable is being opened which can't find this path.  So generally, first is to identify the binary being loaded (exact path) and second is to run that binary in the same way jmeter is to see what's missing.  In [my experience on Linux](https://github.com/Homebrew/homebrew-core/blob/76b5fdef7cfe963c73843d56193f79c0843346b5/Formula/openjdk.rb#L115-L119), it's often just a matter of creating a symlink to the missing file.

**Comment by tresf** [comment]
Some more information about how jmeter launches... It uses the [following command](https://github.com/apache/jmeter/blob/659c1ff5eaea941eb7ad0638b58e904dcc06d961/bin/jmeter#L199):

```
"$JAVA_HOME/bin/java" $ARGS $JVM_ARGS $JMETER_OPTS -jar "$PRGDIR/ApacheJMeter.jar" "$@
```

... which appears to be correct.

The `JAVA_HOME` calculation varies based on [quite a bit of variables](https://github.com/apache/jmeter/blob/master/bin/jmeter), but seems to honor that which is [passed in by the homebrew launcher](https://github.com/Homebrew/homebrew-core/blob/ef5c358418d6299cbbac38b3bce4a8ba8c03d82a/Formula/jmeter.rb#L29).

So assuming there's nothing wrong with the portion `Formula["openjdk"].opt_prefix` on Intel, and there's nothing polluting library loading (e.g. [`DYLD_LIBRARY_PATH`](https://stackoverflow.com/q/3146274/3196753) loading) would expect the same messages to appear for ANY Java application on Intel, no?

**Comment by cyrilhl** [comment]
I using @ajit-y method by using jenv to set java version to adoptopenjdk-11 and edit " /usr/local/Cellar/jmeter/5.4.1/bin/jmeter" to the following without specific a java home also working.
`
#!/bin/bash
exec "/usr/local/Cellar/jmeter/5.4.1/libexec/bin/jmeter"  "$@"
`

**Comment by tudorsss** [comment]
Alright @ajit-y , it works for me now too, what I did was:

`brew install brew install adoptopenjdk` - this installed for me adoptopenjdk-16.jdk
`cd /usr/local/Cellar/jmeter/5.4.1/bin`
`vim jmeter`
`i`
commented out the first two lines with #, like so:

`#!/bin/bash
#JAVA_HOME="/usr/local/opt/openjdk" exec "/usr/local/Cellar/jmeter/5.4.1/libexec/bin/jmeter"  "$@"`


then added the secret sauce:
`
JAVA_HOME="/Library/Java/JavaVirtualMachines/adoptopenjdk-16.jdk/Contents/Home" exec "/usr/local/Cellar/jmeter/5.4.1/libexec/bin/jmeter" "$@"`

then I pressed ESC
then `:wq!`
then the problem got fixed! No more terminal errors.

**Comment by sbibauw** [comment]
For anybody landing here after issues using OpenJDK on **Apple Silicon/M1 Mac**, the solution was simply to use Eclipse **Temurin** (previously AdoptOpenJDK) instead:

```bash
brew install temurin
```

If needed, remove the symlink that you might have created to your keg-only brewed `openjdk` (if you followed the Caveat in `brew info openjdk`):

```bash
# In case you did: sudo ln -sfn /opt/homebrew/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk
rm -i /Library/Java/JavaVirtualMachines/openjdk.jdk
```

(I had an error in R, from `rJava`:

```
.onLoad failed in loadNamespace() for 'rJava', details:
  call: dyn.load(jvm, FALSE)
  error: unable to load shared object '/opt/homebrew/Cellar/openjdk/16.0.2/libexec/openjdk.jdk/Contents/Home/lib/server/libjvm.dylib':
  dlopen(/opt/homebrew/Cellar/openjdk/16.0.2/libexec/openjdk.jdk/Contents/Home/lib/server/libjvm.dylib, 10): no suitable image found.  Did find:
	/opt/homebrew/Cellar/openjdk/16.0.2/libexec/openjdk.jdk/Contents/Home/lib/server/libjvm.dylib: mach-o, but wrong architecture
```
Installing Temurin and removing the symlink to OpenJDK solved it for me.)

**Comment by cho-m** [comment]
I believe `JavaNativeFoundation` was removed when I looked into it in OpenJDK 17 (#85174 - Sept 2021) and the change was also officially backported to OpenJDK 11.0.14.

So, the original messages shouldn't be relevant. If there is still an issue, it will require additional information beyond what is available here.

**Comment by carlocab** [comment]
This should be fixed in #111255. See #111068.
